### PR TITLE
Tramming Wizard

### PR DESCRIPTION
**Pull request by Speaka**
Work in progress! Help and ideas welcome!

### Description
I was tinkering with G35 before and didn't like the workflow very much. Execute command, adjust bed, exec again, see still differences, adjust and so on...
I really like guided adjustment workflows so i decided to develop my own "Bed Tramming Wizard" based on the G35 #16897 command.

### Benefits
What's the wizardry?

1. Call the tool from Menu Advanced
2. Select the Tramming point from which the adjustments should be calculated.
3. Click to probe this Point. (Saving measured Z)
4. Select another point.
5. Click to probe this
6. See the difference in measured z distance (and screw turns)
7. Adjust the bed
8. Click to probe again <- this is the main benefit
Repeat 6-8 until point is levelled sufficiently
9. Select next point and repeat until all is fine

### What needs to be done?

- [x] Clean the code. Correct makros.
- [x] add specific defines 
- [x] sanity checks
- [x] Optimize the menu tree and functions
- [x] ~~Add screw turn adjustment output~~
- [x] debug on other machines and displays
- [x] debug different probe types
- [ ] translation
- [ ] automate most possible steps (Autoprobing)
- [ ] visualize measurements [Datagrid, adjustment bar]

### legal notice ;)
I'm definitely no skilled coder and pretty new to marlin. I did a lot of copy pasta to get this running. It works but still has it flaws and a lot of space for improvements.
I'm filing this Draft PR to get some responses, hints and ideas from other users and the team.
 Let me know what you think about this feature!
- Is it worth investing more time?
- Would you use this?
- Would a beginner benefit from this. Especially with diy printers this would be helpful in my mind.
- Whats missing?
- Whats wrong?

Edit: Didn't see the FRs. Possibly adressing #19480 #18489

### CODE REVIEW

**Comment by ETE-Design** [comment]
Have asked about this feature, cause I don't want to connect a computer to tram it, so would use it :-) Is it made with same GFX like BLTouch Offset?

**Comment by Speaka** [comment]
I don't have BL-Touch installed, so idk the menu. I will have a look for it (and see what features may be stolen from there)

**Comment by ETE-Design** [comment]
@thinkyhead Have just tryed to compile with your changes using UBL, but got this error:
```
Marlin\src\lcd\menu\menu_tramming_wizard.cpp: In function 'bool probe_single_point()':
Marlin\src\lcd\menu\menu_tramming_wizard.cpp:43:77: error: 'selected_point' was not declared in this scope
   const float z_probed_height = probe.probe_at_point(screws_tilt_adjust_pos[selected_point], PROBE_PT_RAISE, 0, true);
                                                                             ^~~~~~~~~~~~~~
Marlin\src\lcd\menu\menu_tramming_wizard.cpp:43:77: note: suggested alternative: 'selectFunc_t'
   const float z_probed_height = probe.probe_at_point(screws_tilt_adjust_pos[selected_point], PROBE_PT_RAISE, 0, true);
                                                                             ^~~~~~~~~~~~~~
                                                                             selectFunc_t
Marlin\src\lcd\menu\menu_tramming_wizard.cpp:45:3: error: 'z_measured' was not declared in this scope
   z_measured[selected_point] = z_probed_height;
   ^~~~~~~~~~
Marlin\src\lcd\menu\menu_tramming_wizard.cpp: In function 'void _menu_single_probe(uint8_t)':
Marlin\src\lcd\menu\menu_tramming_wizard.cpp:50:3: error: 'selected_point' was not declared in this scope
   selected_point = point;
   ^~~~~~~~~~~~~~
Marlin\src\lcd\menu\menu_tramming_wizard.cpp:50:3: note: suggested alternative: 'selectFunc_t'
   selected_point = point;
   ^~~~~~~~~~~~~~
   selectFunc_t
In file included from Marlin\src\lcd\menu\menu_tramming_wizard.cpp:31:0:
Marlin\src\lcd\menu\menu_tramming_wizard.cpp:54:55: error: 'z_measured' was not declared in this scope
   STATIC_ITEM(MSG_LAST_VALUE_SP, SS_LEFT, ftostr42_52(z_measured[0] - z_measured[point])); // Print diff
                                                       ^
Marlin\src\lcd\menu\menu_item.h:314:49: note: in definition of macro 'STATIC_ITEM_INNER_P'
     MenuItem_static::draw(_lcdLineNr, PLABEL, ##V);     \
                                                 ^
Marlin\src\lcd\menu\menu_item.h:345:56: note: in expansion of macro 'STATIC_ITEM_P'
 #define STATIC_ITEM(LABEL, V...)                       STATIC_ITEM_P(GET_TEXT(LABEL), ##V)
                                                        ^~~~~~~~~~~~~
Marlin\src\lcd\menu\menu_tramming_wizard.cpp:54:3: note: in expansion of macro 'STATIC_ITEM'
   STATIC_ITEM(MSG_LAST_VALUE_SP, SS_LEFT, ftostr42_52(z_measured[0] - z_measured[point])); // Print diff
   ^~~~~~~~~~~
Marlin\src\lcd\menu\menu_tramming_wizard.cpp: In function 'void goto_tramming_wizard()':
Marlin\src\lcd\menu\menu_tramming_wizard.cpp:81:3: error: 'selected_point' was not declared in this scope
   selected_point = 0;
   ^~~~~~~~~~~~~~
   selected_point = 0;
   ^~~~~~~~~~~~~~
   selectFunc_t
```

**Comment by qwewer0** [comment]
Is there a way to add movement that take `BLTOUCH_HS_MODE` into consideration? Like it is in G35.cpp: https://github.com/MarlinFirmware/Marlin/blob/f131f58124c6f720ed6e9004ce0aa48cebb28aac/Marlin/src/gcode/bedlevel/G35.cpp#L127
And not sure if it is intended or not, but BLTouch pin stays deployed through the whole process with `BLTOUCH_HS_MODE` enabled.

**Comment by thisiskeithb** [comment]
> And not sure if it is intended or not, but BLTouch pin stays deployed through the whole process with BLTOUCH_HS_MODE enabled.

That's how `BLTOUCH_HS_MODE` works. From a few lines up:
https://github.com/MarlinFirmware/Marlin/blob/f131f58124c6f720ed6e9004ce0aa48cebb28aac/Marlin/src/gcode/bedlevel/G35.cpp#L124

**Comment by qwewer0** [comment]
@thisiskeithb I'm not talking about `ASSISTED_TRAMMING`, for that it is obvious that it moves with pin deployed, but then it stows after everything is done, so by that, in my understanding it should stow after each measurement in `ASSISTED_TRAMMING_WIZARD` just like without `BLTOUCH_HS_MODE`, because if it would timeout to the main screen then it will leave the pin deployed which could be dangerous.

**Comment by qwewer0** [comment]
I hope that I'm not wrong, and by adding 
```
current_position.z = (Z_CLEARANCE_BETWEEN_PROBES) + (7 * ENABLED(BLTOUCH_HS_MODE));
```
before
https://github.com/MarlinFirmware/Marlin/blob/bcdbe207dfa5abd87a0c6f1449be829830025997/Marlin/src/lcd/menu/menu_tramming.cpp#L46

`ASSISTED_TRAMMING_WIZARD` will move to the point to be measured like `ASSISTED_TRAMMING` does. (Raising Z axis for `BLTOUCH_HS_MODE` before XY movement)

Don't know how to contribute to or edit others PR, so this is the best I can do to help.

**Comment by Speaka** [comment]
It is in probe.cpp [here ](https://github.com/MarlinFirmware/Marlin/blob/ba2939b1dc55c04f879ce96e9af990287b4468d2/Marlin/src/module/probe.cpp#L702-L703)
```
else if (raise_after == PROBE_PT_STOW)
      if (stow()) measured_z = NAN;   // Error on stow?
```
So calling `probe_at_point()` should forward `PROBE_PT_STOW` instead of `PROBE_PT_RAISE` with bltouch and it will stow the pin. Correct?

**Comment by ETE-Design** [comment]
@thinkyhead Seems to be working as expected now 👍

File: Marlin/src/lcd/menu/menu_tramming.cpp [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+}
+
+void tramming_wizard_menu() {
+  DEBUG_ECHOLNPAIR("Screen: tramming_wizard_menu");
+  START_MENU();
+  STATIC_ITEM(MSG_SELECT_ORIGIN);
+
+  // Draw a menu item for each tramming point
+  LOOP_L_N(i, G35_PROBE_COUNT)
+    SUBMENU_N_P(i, tramming_point_name[i], []{ _menu_single_probe(MenuItemBase::itemIndex); });
```
Comment (Speaka) [inline Marlin/src/lcd/menu/menu_tramming.cpp]: The menu doesn't draw the point names on my 8bit. Am I missing something or is this array empty?

Comment (thinkyhead) [reply Marlin/src/lcd/menu/menu_tramming.cpp]: Maybe add `&` in front of `tramming_point_name[i]` ?

Comment (Speaka) [reply Marlin/src/lcd/menu/menu_tramming.cpp]: Not enough. Same Problem like this i think: #19926
Pushed changes. Works for me.
Thanks for the awesome help by the way!

Comment (thinkyhead) [reply Marlin/src/lcd/menu/menu_tramming.cpp]: Happy to help. Thanks for bringing the improvements!

**Comment by hamster65** [comment]
Feature request: If not already implemented, an option to move the print head out of the way while the user turns the screw would be nice. With my setup, I cannot reach the screws of the bed when the print head is above them....

**Comment by Speaka** [comment]
Added a PR for this. @hamster65
Please try if thius suits your needs.

**Comment by hamster65** [comment]
@Speaka Great, thanks! Looks good to me but I cannot test yet because of https://github.com/MarlinFirmware/Marlin/issues/19732

**Comment by ETE-Design** [comment]
@thinkyhead Funny thing, before you merged it into the Bugfix I had No problem with the memory required even with All the other features I needed activated, and it was merged up with the almost new bugfix... Now when merged into Bugfix I don't have enough memory for it anymore :-( Have already activated Slim Menu and so...

**Pull request merged**